[Vision] add early stopping strategies

zhreshold opened the pull request:
*Issue #, if available:*

*Description of changes:*
Add early stopping control for ImagePredictor and ObjectDetector.

New hyperparameters that controls the early stopping bahavior:

```python
"""
early_stop_patience : int, default=-1
    Number of epochs with no improvement after which train is early stopped. Use -1 to disable.
early_stop_min_delta : float, default=1e-4
    The small delta value to ignore when evaluating the metric. A large delta helps stablize the early
    stopping strategy against tiny fluctuation, e.g. 0.5->0.49->0.48->0.499->0.500001 is still considered as
    a good timing for early stopping.
early_stop_baseline : float, default=0.0
    The minimum(baseline) value to trigger early stopping. For example, with `early_stop_baseline=0.5`,
    early stopping won't be triggered if the metric is less than 0.5 even if plateau is detected.
early_stop_max_value : float, default=1.0
    The max value for metric, early stop training instantly once the max value is achieved.
You can get the list of accepted hyperparameters in `config.yaml` saved by this predictor.
"""
```

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

szha commented:
Job PR-1039-2 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1039/2/index.html

Innixma commented on vision/src/autogluon/vision/detector/detector.py:
```diff
@@ -164,6 +168,17 @@ def fit(self,
+                The small delta value to ignore when evaluating the metric. A large delta helps stablize the early
+                stopping strategy against tiny fluctuation, e.g. 0.5->0.49->0.48->0.499->0.500001 is still considered as
+                a good timing for early stopping.
+            early_stop_baseline : float, default=0.0
```
Should we have these disabled (None) by default? This seems metric specific. RMSE / r2 / AUC / custom metrics may differ in min/max values.

Innixma commented on vision/src/autogluon/vision/detector/detector.py:
```diff
@@ -164,6 +168,17 @@ def fit(self,
                 Mini batch size
             lr : float
                 Trainer learning rate for optimization process.
+            early_stop_patience : int, default=-1
```
Should we standardize that None is the way to disable? Since `-1` doesn't make sense for things like `early_stop_baseline` since metrics like r2 can have negative values.

Innixma commented on vision/src/autogluon/vision/predictor/predictor.py:
```diff
@@ -107,7 +107,8 @@ def fit(self,
                         'batch_size': Categorical(8, 16, 32, 64, 128),
-                        'epochs': 200
+                        'epochs': 200,
+                        'early_stop_patience': -1
```
Are we sure this is useful? Why not have it be 50? Is it realistic that it would take 100+ epochs to show improvement?

zhreshold replied:
changed best_quality to use 50 epochs, it's very reasonable for now but somewhat suboptimal comparing to `as-is` because currently the learning rate scheduler is not affected by the early stopping, in the future a LRSchedulerOnPlateau or similar stuff can offer better tradeoffs

Innixma reviewed (approved):
Looks good!

sxjscience reviewed (approved):
LGTM!

szha commented:
Job PR-1039-4 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1039/4/index.html

zhreshold commented:
@Innixma Addressed the comments, after merging this it should improve the overall training time by a large margin especially when the dataset is not that hard.

Innixma commented:
Awesome! Feel free to merge whenever you are ready.
